### PR TITLE
SUMIFS multiple dimensions

### LINKED ISSUE #2041: SUMIF doesn't consider only first column
## Read and complete the full issue template

Do not randomly delete sections. They are here for a reason.

**Do you want to request a *feature* or report a *bug*?**
- [x] Bug
- [ ] Feature
- [ ] Question

**Did you test against the latest CI build?**
- [x] Yes
- [ ] No

If you answered `No`, please test with [the latest development build](https://ci.appveyor.com/project/ClosedXML/ClosedXML/branch/develop/artifacts) first.

**Version of ClosedXML**

0.100.3
(Installed via nuget.org using NuGet package manager)

**What is the current behavior?**

When reading a complex file with lots of formulas and links, it gives wrong formula calculation results

**What is the expected behavior or new feature?**

Complete this.

**Is this a regression from the previous version?**

No

## Reproducibility
**This is an important section. Read it carefully. Failure to do so will cause a 'RTFM' comment.**

No special actions are needed: just open the attached file and take the value of cell (13, 9) from the first sheet as an example

Value in file:
image

The value produced by the code:
image

**Code to reproduce problem:**
```c#
using (var fs = new FileStream(@"E:\Temp\Расчёт гидравлики СНТ Вишня.xlsx", FileMode.Open, FileAccess.Read))
{
    var workbook = new XLWorkbook(fs);
    var worksheet = workbook.Worksheet(1);
    var xlCell = worksheet.Cell(13, 9);
    xlCell.InvalidateFormula();
    MessageBox.Show(xlCell.Value.ToString());
}
```

- [ ] I attached a sample spreadsheet.  (You can drag files on to this issue)
[Расчёт гидравлики СНТ Вишня.xlsx](https://github.com/ClosedXML/ClosedXML/files/11112243/default.xlsx)

### PR DESCRIPTION
**Pull request by igitur**
Fixes #2041

Add support and applicable errors for input ranges having multilpe dimensions (more than 1 column)

### CODE REVIEW

File: ClosedXML/Excel/CalcEngine/CalcEngineHelpers.cs [hunk 1]
```diff
@@ -134,12 +134,26 @@ internal static bool ValueIsBlank(object value)
...
+            return (long)columnCount * (long)rowCount;
+        }
+
+        /// <summary>
+        /// Get dimensions of the specified range without initializing them all
+        /// (which might cause serious performance issues on column-wide calculations).
+        /// </summary>
+        /// <param name="rangeExpression">Expression referring to the cell range.</param>
+        /// <returns>A tuple of column and row counts.</returns>
+        internal static (int columnCount, int rowCount) GetRangeDimensions(XObjectExpression rangeExpression)
```
Comment (jahav) [inline ClosedXML/Excel/CalcEngine/CalcEngineHelpers.cs]: Names of properties of named tuples should be in PascalCase.

File: ClosedXML.Tests/Excel/CalcEngine/MathTrigTests.cs [hunk 2]
```diff
@@ -2063,6 +2063,17 @@ public void SumIfs_ReturnsCorrectValues_ReferenceExampleForSumIf1FromMicrosoft(i
             }
         }
 
+        [Test]
```
Comment (jahav) [inline ClosedXML.Tests/Excel/CalcEngine/MathTrigTests.cs]: `[Test]` is not required if there is a `[TestCase]`

File: ClosedXML.Tests/Excel/CalcEngine/MathTrigTests.cs [hunk 3]
```diff
@@ -1783,6 +1782,26 @@ public void SumDateTimeAndNumber()
...
```
Comment (jahav) [inline ClosedXML.Tests/Excel/CalcEngine/MathTrigTests.cs]: `[Test]` is not required if there is a `[TestCase]`

**Review by jahav — APPROVED** [review]
Looks good. Thanks for investigating, good work. 

I wonder about other functions, SUMIF has a sibling functions AVERAGEIF/S and COUNTIF/S. We don't implement AVERAGEIF, but COUNTIF has a similar problem.

I think that rather than fixing old range functions, it would be better to reimplement them. The `XObjectExpression` adapter works, but I would like to remove it. 

It would require a new method to validate criteria (outlined [here](https://github.com/ClosedXML/ClosedXML/pull/736#issuecomment-1507067695)).

**Comment by igitur** [comment]
> I wonder about other functions, SUMIF has a sibling functions AVERAGEIF/S and COUNTIF/S. We don't implement AVERAGEIF, but COUNTIF has a similar problem.

I can have a look at the sibling functions and fix them as a short-term goal.

 
> I think that rather than fixing old range functions, it would be better to reimplement them. The `XObjectExpression` adapter works, but I would like to remove it.
> 
> It would require a new method to validate criteria (outlined [here](https://github.com/ClosedXML/ClosedXML/pull/736#issuecomment-1507067695)).

Don't think I have the capacity for that now, sorry. Hence the goal of this PR is a short term fix.

**Pull request merged**